src/Access/Common/AuthenticationData.cpp

```diff
@@ -85,6 +95,40 @@ AuthenticationData::Digest AuthenticationData::Util::encodeSHA256(std::string_vi
 #endif
 }
 
+AuthenticationData::Digest AuthenticationData::Util::encodeBcrypt(std::string_view text [[maybe_unused]])
+{
+#if USE_BCRYPT
+    char salt[BCRYPT_HASHSIZE];
+    Digest hash;
+    hash.resize(64);
+    int ret = bcrypt_gensalt(12, salt);
+    if (ret != 0)
+        throw DB::Exception(DB::ErrorCodes::LOGICAL_ERROR, "BCrypt library failed: bcrypt_gensalt returned {}", ret);
+    ret = bcrypt_hashpw(text.data(), salt, reinterpret_cast<char *>(hash.data()));
+    if (ret != 0)
+        throw DB::Exception(DB::ErrorCodes::LOGICAL_ERROR, "BCrypt library failed: bcrypt_hashpw returned {}", ret);
+    return hash;
+#else
+    throw DB::Exception(
+        "bcrypt passwords support is disabled, because ClickHouse was built without bcrypt library",
+        DB::ErrorCodes::SUPPORT_IS_DISABLED);
+#endif
+}
+
+bool AuthenticationData::Util::checkPasswordBcrypt(std::string_view password [[maybe_unused]], const Digest & password_bcrypt [[maybe_unused]])
```

Comment:
Can't `checkPasswordBcrypt()` be implemented like this:

```
bool checkPasswordBcrypt(td::string_view password, const Digest & password_bcrypt)
{
   return encodeBcrypt(password) == password_bcrypt;
}
```

Reply:
We would need to store the salt separately, and the computed hash already contains salt, so we would just store it twice. So we can implement it this way, but I'm not sure why we would want to do it.

And also the function `bcrypt_checkpw` that we use to check the password uses `timing_safe_strcmp` for string comparison to prevent timing attacks, which is also probably good to have.